Add fast path of getting arguments.length for LLInt and Baseline

hyjorc1 opened the pull request:
#### cacf00c5fdeedd4db49495feaa674396691def67
<pre>
Add fast path of getting arguments.length for LLInt and Baseline
<a href="https://bugs.webkit.org/show_bug.cgi?id=261543">https://bugs.webkit.org/show_bug.cgi?id=261543</a>
rdar://115462355

Reviewed by Yusuke Suzuki.

We observed that `arguments.length` is frequently used for reading only
without accessing/modifying arguments self or its other properties. If that&apos;s
the case we should directly get `argumentCountIncludingThis` from CallFrame
without materializing `arguments` object first. So far, DFG and FTL can
handle this well in varargs forwarding phase. This patch adds the fast
path of getting arguments.length for LLInt and Baseline with simple
heuristic in parser layer.

JavaScriptCode:

    function test() {
        return arguments.length === 3;
    }

Bytecode Before:

    Predecessors: [ ]
    [   0] enter
    [   1] create_direct_arguments dst:loc5
    [   3] mov                dst:loc6, src:loc5
    [   6] get_by_id          dst:loc7, base:loc6, property:0, valueProfile:1
    [  12] stricteq           dst:loc7, lhs:loc7, rhs:Int32: 3(const0)
    [  16] ret                value:loc7
    Successors: [ ]

Bytecode After:

    Predecessors: [ ]
    [   0] enter
    [   1] argument_count     dst:loc5
    [   3] stricteq           dst:loc5, lhs:loc5, rhs:Int32: 3(const0)
    [   7] ret                value:loc5
    Successors: [ ]

Micro Benchmark Result:

                                without                     with

get-arguments-length-fast    2.6024+-0.0750     ^      2.4034+-0.0359        ^ definitely 1.0828x faster

* JSTests/microbenchmarks/get-arguments-length-fast.js: Added.
(test):
* JSTests/stress/get-arguments-length-fast.js: Added.
(shouldBe):
(argumentsWithNestedNormalFunction.nested):
(argumentsWithNestedNormalFunction):
(argumentsNonLengthAccess):
(argumentsLengthAssignment):
(argumentsAssignmentLHS):
(argumentsAssignmentRHS):
(argumentsDestructuringAssignment):
(declareVarArguments):
(declareVarArgumentsWithBlock):
(declareLexicalArguments):
(declareLexicalArgumentsWithBlock):
(declareArrowFunctionArguments):
(declareArrowFunctionArgumentsWithBlock):
(declareFunctionArguments.arguments):
(declareFunctionArguments):
(declareFunctionArgumentsWithBlock.arguments):
(declareFunctionArgumentsWithBlock):
(declareParameterArguments):
(returnArguments):
(helper):
(passArgumentsToFunctionCall):
(ifArguments):
(forInArguments):
(forInVarArguments):
(forInLetArguments):
(forOfArguments):
(forOfVarArguments):
(forOfLetArguments):
(withArguments):
(argumentsWithNestedFunctions.nestedNormalFunction):
(argumentsWithNestedFunctions):
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::BytecodeGenerator):
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
(JSC::BytecodeGenerator::useGetArgumentsLengthFast):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::DotAccessorNode::emitBytecode):
* Source/JavaScriptCore/parser/ASTBuilder.h:
(JSC::ASTBuilder::createDotAccess):
(JSC::ASTBuilder::createAssignment):
(JSC::ASTBuilder::createAssignmentElement):
(JSC::ASTBuilder::checkAssignmentForArguments):
(JSC::ASTBuilder::useArgumentsOnlyForReadingLength):
(JSC::ASTBuilder::usesArguments):
* Source/JavaScriptCore/parser/Nodes.h:
(JSC::ExpressionNode::isArgumentsAccess const):
(JSC::ExpressionNode::isArgumentsLengthAccess const):
(JSC::ExpressionNode::isArguments const):
(JSC::ScopeNode::useGetArgumentsLengthFast const):
* Source/JavaScriptCore/parser/Parser.cpp:
(JSC::Parser&lt;LexerType&gt;::parseInner):
* Source/JavaScriptCore/parser/Parser.h:
(JSC::Scope::hasDeclaredArguments const):
(JSC::Scope::setHasDeclaredArguments):
(JSC::Scope::declareCallee):
(JSC::Scope::declareVariable):
(JSC::Scope::declareFunction):
(JSC::Scope::declareLexicalVariable):
(JSC::Scope::addDeclaredVariable):
(JSC::Scope::addDeclareLexicalVariable):
(JSC::Scope::addDeclaredParameters):
(JSC::Scope::declareParameter):
(JSC::Scope::getSloppyModeHoistedFunctions):
(JSC::Scope::setReadArgumentsLengthOnly):
(JSC::Scope::useGetArgumentsLengthFast const):
(JSC::Parser::popScopeInternal):
* Source/JavaScriptCore/parser/ParserModes.h:
* Source/JavaScriptCore/parser/VariableEnvironment.h:

Canonical link: <a href="https://commits.webkit.org/268480@main">https://commits.webkit.org/268480@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3ce8bb6054e4566b2f6502d08004d4990708bf2b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19227 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19642 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20240 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21116 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17990 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22913 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19768 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19660 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19448 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19510 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16717 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21986 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16700 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17503 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23856 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/16687 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17758 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17678 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21815 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/18554 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18268 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15470 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/22617 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17397 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/5602 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4740 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21754 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/23867 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18103 "Built successfully") | | [❌ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5338 "Found 1 jsc stress test failure: microbenchmarks/array-from-object-func.js.lockdown") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on JSTests/microbenchmarks/get-arguments-length-fast.js:
```diff
@@ -0,0 +1,7 @@
+function test() {
+    return arguments.length === 3;
+}
```
Let's set noInline and measure performance

MenloDorian commented on Source/JavaScriptCore/parser/Nodes.h:
```diff
@@ -906,6 +910,18 @@ namespace JSC {
+        bool isArgumentsLengthAccess(VM& vm) const final
+        {
+            return m_base->isResolveNode()
+                && static_cast<ResolveNode*>(m_base)->identifier() == vm.propertyNames->arguments
```
Why does this need to check `m_base->isResolveNode()` but `isArgumentsAccess` does not need to do the same check?  Why not just call `m_base->isArguments(vm)` instead?  I see that `isArguments` is a virtual function though.

hyjorc1 replied:
Yes, we should use `m_base->isArguments(vm)`.

Constellation commented on Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
```diff
@@ -1234,6 +1234,8 @@ namespace JSC {
         void pushPrivateAccessNames(const PrivateNameEnvironment*);
         void popPrivateAccessNames();
 
+        bool shouldUseGetArgumentsLengthFast() { return m_shouldUseGetArgumentsLengthFast; };
```
Make const member function

Constellation commented on Source/JavaScriptCore/parser/ASTBuilder.h:
```diff
@@ -1106,8 +1111,17 @@ class ASTBuilder {
+        bool m_usesArgumentsLength { false };
     };
 
+    void checkArgumentsLengthModification(const ExpressionNode* node)
+    {
+        // Since we exclude pattern `arguments.length` to enable ArgumentsFeature,
+        // we need re-enable ArgumentsFeature for `arguments.length` modification.
+        if (node && node->isArgumentsLengthAccess(m_vm))
+            usesArguments();
+    }
```
I don't think this is the right approach. We should disable `arguments.length` if we found `arguments` identifier resolves regardless of `arguments.xxx = ` assignment. If it has a getter, then it can do whatever we want.

hyjorc1 replied:
We need this for these cases:

```
function argumentsLengthAssignment() {
    arguments.length = 3;
    shouldBe(arguments.length, 3);
}
argumentsLengthAssignment(1);
```
and
```
function argumentsLengthIncAndDec() {
    arguments.length++;
    arguments.length--;
    ++arguments.length;
    --arguments.length;
    shouldBe(arguments.length, 0);
}
argumentsLengthIncAndDec();
```

Constellation commented on Source/JavaScriptCore/parser/ASTBuilder.h:
```diff
@@ -197,9 +197,9 @@ class ASTBuilder {
     bool isNewTarget(ExpressionNode* node) { return node->isNewTarget(); }
     bool isImportMeta(ExpressionNode* node) { return node->isImportMeta(); }
-    ExpressionNode* createResolve(const JSTokenLocation& location, const Identifier& ident, const JSTextPosition& start, const JSTextPosition& end)
+    ExpressionNode* createResolve(const JSTokenLocation& location, const Identifier& ident, const JSTextPosition& start, const JSTextPosition& end, const bool needToCheckUsesArguments = true)
```
`needToCheckUsesArguments` should be `false` by default. We should mark `usesArguments` by default, and allowing to skip it only when we are focusing on `arguments.length`.

hyjorc1 replied:
`needToCheckUsesArguments` should be `true` by default since

```
    ExpressionNode* createResolve(const JSTokenLocation& location, const Identifier& ident, const JSTextPosition& start, const JSTextPosition& end, const bool needToCheckUsesArguments = true)
    {
        if (needToCheckUsesArguments && m_vm.propertyNames->arguments == ident)
            usesArguments();
        ...
```

Constellation commented on Source/JavaScriptCore/parser/ASTBuilder.h:
```diff
@@ -281,7 +281,8 @@ class ASTBuilder {
             usesSuperProperty();
-        
+        if (base->isArguments(m_vm) && *property == m_vm.propertyNames->length)
+            m_scope.m_usesArgumentsLength = true;
```
I don't think we should do this. Remove it.
What we should do is marking `arguments` resolve by default, and not marking usesArguments only when we saw `arguments.length`. The current approach is very fragile and dangerous.

Constellation commented on Source/JavaScriptCore/parser/ASTBuilder.h:
```diff
@@ -981,6 +982,7 @@ class ASTBuilder {
     ExpressionNode* createAssignment(const JSTokenLocation& location, int& assignmentStackDepth, ExpressionNode* rhs, int initialAssignmentCount, int currentAssignmentCount, const JSTextPosition& lastTokenEnd)
     {
         AssignmentInfo& info = m_assignmentInfoStack.last();
+        checkArgumentsLengthModification(info.m_node);
```
Ditto. Remove this.

hyjorc1 replied:
We need this.

Constellation commented on Source/JavaScriptCore/parser/ASTBuilder.h:
```diff
@@ -1066,6 +1068,7 @@ class ASTBuilder {
 
     AssignmentElement createAssignmentElement(const Expression& assignmentTarget, const JSTextPosition& start, const JSTextPosition& end)
     {
+        checkArgumentsLengthModification(assignmentTarget);
```
Ditto.

hyjorc1 replied:
We need this.

Constellation commented on Source/JavaScriptCore/parser/ASTBuilder.h:
```diff
@@ -1658,11 +1672,13 @@ ExpressionNode* ASTBuilder::makeAssignNode(const JSTokenLocation& location, Expr
 
 ExpressionNode* ASTBuilder::makePrefixNode(const JSTokenLocation& location, ExpressionNode* expr, Operator op, const JSTextPosition& start, const JSTextPosition& divot, const JSTextPosition& end)
 {
+    checkArgumentsLengthModification(expr);
```
Ditto.

hyjorc1 replied:
We need this.

Constellation commented on Source/JavaScriptCore/parser/ASTBuilder.h:
```diff
@@ -1658,11 +1672,13 @@ ExpressionNode* ASTBuilder::makeAssignNode(const JSTokenLocation& location, Expr
 
 ExpressionNode* ASTBuilder::makePostfixNode(const JSTokenLocation& location, ExpressionNode* expr, Operator op, const JSTextPosition& start, const JSTextPosition& divot, const JSTextPosition& end)
 {
+    checkArgumentsLengthModification(expr);
```
Ditto.

hyjorc1 replied:
We need this.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -5017,11 +5032,11 @@ template <class TreeBuilder> typename TreeBuilder::TemplateLiteral Parser<LexerT
 
 template <class LexerType>
-template <class TreeBuilder> TreeExpression Parser<LexerType>::createResolveAndUseVariable(TreeBuilder& context, const Identifier* ident, bool isEval, const JSTextPosition& start, const JSTokenLocation& location)
+template <class TreeBuilder> TreeExpression Parser<LexerType>::createResolveAndUseVariable(TreeBuilder& context, const Identifier* ident, bool isEval, const JSTextPosition& start, const JSTokenLocation& location, const bool needToCheckUsesArguments)
```
I don't think this is the right approach as I said.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -5090,7 +5105,24 @@ template <class TreeBuilder> TreeExpression Parser<LexerType>::parsePrimaryExpre
+                needToCheckUsesArguments = match(DOT) || match(OPENBRACKET);
+                restoreSavePoint(context, argumentsDotLengthDotSavePoint);
+            }
+            restoreSavePoint(context, argumentsDotLengthSavePoint);
```
Restoring twice looks wrong.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -5090,7 +5105,24 @@ template <class TreeBuilder> TreeExpression Parser<LexerType>::parsePrimaryExpre
-        return createResolveAndUseVariable(context, ident, *ident == m_vm.propertyNames->eval, start, location);
+        // If encounter JavaScript code `arguments.length`, a fast path can be applied to get
+        // `arguments.length` directly from CallFrame without materializing `arguments` object
+        // first. However, we should avoid that when parsing array fuctnion or other property accesses.
```
I don't think this is the right place to insert these code. Why not adding these kind of things into parseMemberExpression as I said?

Constellation reviewed (changes requested):
Commented

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -5024,6 +5026,60 @@ template <class TreeBuilder> TreeExpression Parser<LexerType>::createResolveAndU
+            next();
+            if (match(DOT) || match(OPENBRACKET))
+                goto failToFindArgumentsDotLengthPattern;
+            restoreSavePoint(context, argumentsDotLengthSavePoint);
```
Why? I think what we should do is just creating an argument ResolveNode without marking usesArguments. That's it.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -5352,6 +5408,8 @@ template <class TreeBuilder> TreeExpression Parser<LexerType>::parseMemberExpres
         const bool isAsync = matchContextualKeyword(m_vm.propertyNames->async);
 
+        if (TreeExpression argumentsDotLengthExpression = tryParseArgumentsDotLengthForFastPath(context, startLocation, expressionStart))
+            return argumentsDotLengthExpression;
```
Why do we return here? Why not just return ResolveNode with `arguments` without marking `usesArguments` and use the rest of the code as is?

Constellation reviewed (changes requested):
Let's also add a test,

```
arguments.length()
```

The reason is that, in the above case, we need to get `arguments` as a |this|.
But actually, since length is number, we do not need that. So what we should do is probably just using `undefined` as a base in this case (so we can avoid creating `arguments`).

Constellation reviewed (approved):
r=me

hyjorc1 committed:
Add fast path of getting arguments.length for LLInt and Baseline
https://bugs.webkit.org/show_bug.cgi?id=261543
rdar://115462355

Reviewed by Yusuke Suzuki.

We observed that `arguments.length` is frequently used for reading only
without accessing/modifying arguments self or its other properties. If that's
the case we should directly get `argumentCountIncludingThis` from CallFrame
without materializing `arguments` object first. So far, DFG and FTL can
handle this well in varargs forwarding phase. This patch adds the fast
path of getting arguments.length for LLInt and Baseline with simple
heuristic in parser layer.

JavaScriptCode:

    function test() {
        return arguments.length === 3;
    }

Bytecode Before:

    Predecessors: [ ]
    [   0] enter
    [   1] create_direct_arguments dst:loc5
    [   3] mov                dst:loc6, src:loc5
    [   6] get_by_id          dst:loc7, base:loc6, property:0, valueProfile:1
    [  12] stricteq           dst:loc7, lhs:loc7, rhs:Int32: 3(const0)
    [  16] ret                value:loc7
    Successors: [ ]

Bytecode After:

    Predecessors: [ ]
    [   0] enter
    [   1] argument_count     dst:loc5
    [   3] stricteq           dst:loc5, lhs:loc5, rhs:Int32: 3(const0)
    [   7] ret                value:loc5
    Successors: [ ]

Micro Benchmark Result:

                                without                     with

get-arguments-length-fast    2.6024+-0.0750     ^      2.4034+-0.0359        ^ definitely 1.0828x faster

* JSTests/microbenchmarks/get-arguments-length-fast.js: Added.
(test):
* JSTests/stress/get-arguments-length-fast.js: Added.
(shouldBe):
(argumentsWithNestedNormalFunction.nested):
(argumentsWithNestedNormalFunction):
(argumentsNonLengthAccess):
(argumentsLengthAssignment):
(argumentsAssignmentLHS):
(argumentsAssignmentRHS):
(argumentsDestructuringAssignment):
(declareVarArguments):
(declareVarArgumentsWithBlock):
(declareLexicalArguments):
(declareLexicalArgumentsWithBlock):
(declareArrowFunctionArguments):
(declareArrowFunctionArgumentsWithBlock):
(declareFunctionArguments.arguments):
(declareFunctionArguments):
(declareFunctionArgumentsWithBlock.arguments):
(declareFunctionArgumentsWithBlock):
(declareParameterArguments):
(returnArguments):
(helper):
(passArgumentsToFunctionCall):
(ifArguments):
(forInArguments):
(forInVarArguments):
(forInLetArguments):
(forOfArguments):
(forOfVarArguments):
(forOfLetArguments):
(withArguments):
(argumentsWithNestedFunctions.nestedNormalFunction):
(argumentsWithNestedFunctions):
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::BytecodeGenerator):
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
(JSC::BytecodeGenerator::useGetArgumentsLengthFast):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::DotAccessorNode::emitBytecode):
* Source/JavaScriptCore/parser/ASTBuilder.h:
(JSC::ASTBuilder::createDotAccess):
(JSC::ASTBuilder::createAssignment):
(JSC::ASTBuilder::createAssignmentElement):
(JSC::ASTBuilder::checkAssignmentForArguments):
(JSC::ASTBuilder::useArgumentsOnlyForReadingLength):
(JSC::ASTBuilder::usesArguments):
* Source/JavaScriptCore/parser/Nodes.h:
(JSC::ExpressionNode::isArgumentsAccess const):
(JSC::ExpressionNode::isArgumentsLengthAccess const):
(JSC::ExpressionNode::isArguments const):
(JSC::ScopeNode::useGetArgumentsLengthFast const):
* Source/JavaScriptCore/parser/Parser.cpp:
(JSC::Parser<LexerType>::parseInner):
* Source/JavaScriptCore/parser/Parser.h:
(JSC::Scope::hasDeclaredArguments const):
(JSC::Scope::setHasDeclaredArguments):
(JSC::Scope::declareCallee):
(JSC::Scope::declareVariable):
(JSC::Scope::declareFunction):
(JSC::Scope::declareLexicalVariable):
(JSC::Scope::addDeclaredVariable):
(JSC::Scope::addDeclareLexicalVariable):
(JSC::Scope::addDeclaredParameters):
(JSC::Scope::declareParameter):
(JSC::Scope::getSloppyModeHoistedFunctions):
(JSC::Scope::setReadArgumentsLengthOnly):
(JSC::Scope::useGetArgumentsLengthFast const):
(JSC::Parser::popScopeInternal):
* Source/JavaScriptCore/parser/ParserModes.h:
* Source/JavaScriptCore/parser/VariableEnvironment.h:

Canonical link: https://commits.webkit.org/268480@main
